cmd/util/cmd/execution-state-extract/execution_state_extract.go

```diff
@@ -191,3 +192,167 @@ func writeStatusFile(fileName string, e error) error {
 	err := os.WriteFile(fileName, checkpointStatusJson, 0644)
 	return err
 }
+
+func extractExecutionStateFromPayloads(
+	log zerolog.Logger,
+	dir string,
+	outputDir string,
+	nWorker int, // number of concurrent worker to migation payloads
+	runMigrations bool,
+	inputPayloadFile string,
+	outputPayloadFile string,
+	exportPayloadsByAddresses []common.Address,
+) error {
+
+	payloads, err := util.ReadPayloadFile(log, inputPayloadFile)
+	if err != nil {
+		return err
+	}
+
+	log.Info().Msgf("read %d payloads", len(payloads))
+
+	migrations := newMigrations(log, dir, nWorker, runMigrations)
+
+	payloads, err = migratePayloads(log, payloads, migrations)
+	if err != nil {
+		return err
+	}
+
+	exportPayloads := len(outputPayloadFile) > 0
+	if exportPayloads {
+		exportedPayloadCount, err := util.CreatePayloadFile(
+			log,
+			outputPayloadFile,
+			payloads,
+			exportPayloadsByAddresses,
+		)
+		if err != nil {
+			return fmt.Errorf("cannot generate payloads file: %w", err)
+		}
+
+		log.Info().Msgf("Exported %d payloads out of %d payloads", exportedPayloadCount, len(payloads))
```

Comment:
We break the state extraction into 2 steps, the first step to output all payloads into a file after the first migration, and the second step to read all the payloads from the output file and continue running the remaining migrations and extract state. 

I wonder how can we make the process verifiable in case someone is trying to reproduce the same process, and try to validate each intermediate steps? Basically we need a way to guarantee and prove that the payloads did not change when reading it back and continue with the remaining migrations and state extraction.

One idea I had is to compute some hash (or checksum) of all the payloads before writing them to disk, and validate it after reading and decode the payloads. 

Thoughts?

Reply:
> I wonder how can we make the process verifiable in case someone is trying to reproduce the same process, and try to validate each intermediate steps? Basically we need a way to guarantee and prove that the payloads did not change when reading it back and continue with the remaining migrations and state extraction.

Making this verifiable "in case someone is trying to reproduce the same process" is not the [primary use case now](https://discord.com/channels/613813861610684416/1207310383602335905/1207373538613268530) (this PR is to help speed up development/testing) but it makes sense to add a hash for other use cases. :+1: 

> One idea I had is to compute some hash (or checksum) of all the payloads before writing them to disk, and validate it after reading and decode the payloads.
>
> Thoughts?

Due to parallelism, the payloads within the output file may not be in the same sequence each time this program is used.  In practice payload file would still produce the same checkpoint when used as input (as long as sequence of payloads inside is the only thing that changed and the contents of all payloads are unchanged).

I'll merge this PR and open a separate PR to address the "reproduce the same process" use case by:
- making the payloads have deterministic sequence in the output payload file
- computing and adding a hash to the payload file representing all the payloads